File: python/z2/file_transport_plugin.py
from __future__ import annotations

import asyncio
import json
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Protocol

from .transport_plugin import OpenAPIClientTransportPlugin, TransportRequest, TransportResult


class _FileRuntime(Protocol):
    _file_poll_interval: float

    def next_rpc_id(self) -> str:
        ...

    def _resolve_file_queue_paths(self) -> tuple[Path, Path]:
        ...

    def _coerce_input_model(self, input_data: Any) -> Any:
        ...


@dataclass
class _FileConnection:
    operation: Any
    request_id: str
    response_path: Path
    events_path: Path
    seen_events: int = 0


def create_file_transport_plugin(runtime: _FileRuntime, operation: Any) -> OpenAPIClientTransportPlugin:
    class FileTransportPlugin:
        name = "file"

        def can_handle(self, request: dict[str, str]) -> bool:
            return request.get("transport_mode") == "file"

        def connect(self, request: TransportRequest) -> _FileConnection:
            inbox, outbox = runtime._resolve_file_queue_paths()
            inbox.mkdir(parents=True, exist_ok=True)
            outbox.mkdir(parents=True, exist_ok=True)
            request_id = f"file-{int(time.time() * 1000)}-{runtime.next_rpc_id()}"
            return _FileConnection(
                operation=operation,
                request_id=request_id,
                response_path=outbox / f"{request_id}.response.json",
                events_path=outbox / f"{request_id}.events.jsonl",
            )

        def send_request(self, connection: _FileConnection, request: TransportRequest) -> None:
            inbox, _ = runtime._resolve_file_queue_paths()
            request_path = inbox / f"{connection.request_id}.json"
            request_path.write_text(
                json.dumps({
                    "id": connection.request_id,
                    "operationId": connection.operation.operation_id,
                    "method": connection.operation.method,
                    "path": connection.operation.path,
                    "headers": request.headers,
                    "input": runtime._coerce_input_model(request.params),
                }),
                encoding="utf-8",
            )

        def get_result(self, connection: _FileConnection, request: TransportRequest) -> Any:
            while True:
                if request.on_event is not None and connection.events_path.exists():
                    lines = connection.events_path.read_text(encoding="utf-8").splitlines()
                    for line in lines[connection.seen_events:]:
                        if line.strip():
                            maybe = request.on_event(json.loads(line))
                            if hasattr(maybe, "__await__"):
                                asyncio.run(maybe)
                    connection.seen_events = len(lines)
                if connection.response_path.exists():
                    response = json.loads(connection.response_path.read_text(encoding="utf-8"))
                    if not response.get("ok"):
                        error = response.get("error") or {}
                        return TransportResult(id=request.id, ok=False, error=RuntimeError(error.get("message", "File queue request failed")))
                    return TransportResult(id=request.id, ok=True, result=response.get("result"))
                time.sleep(runtime._file_poll_interval)

    return FileTransportPlugin()

File: typescript/src/client/file-transport-plugin.ts
import type {
  OpenAPIClientTransportOutcome,
  OpenAPIClientTransportPlugin,
  OpenAPIClientTransportRequest,
} from './transport-plugin'

interface FileTransportRuntime {
  nextRequestId(prefix: string): string
  stringifyHeaders(headers: Record<string, string | number | boolean | undefined>): Record<string, string>
  parseJson(text: string): unknown
  delay(ms: number): Promise<void>
  fileQueue?: {
    resolvePaths(): Promise<{ inbox: string; outbox: string }> | { inbox: string; outbox: string }
    pollIntervalMs: number
    mkdir(path: string): Promise<void>
    write(path: string, content: string): Promise<void>
    read(path: string): Promise<string>
  }
}

interface FileConnection {
  responsePath: string
  eventsPath: string
  seenEvents: number
}

export function createFileTransportPlugin(runtime: FileTransportRuntime): OpenAPIClientTransportPlugin<FileConnection> {
  return {
    name: 'file',
    canHandle: ({ transportMode }) => transportMode === 'file',
    async connect(request: OpenAPIClientTransportRequest): Promise<FileConnection> {
      if (!runtime.fileQueue) {
        throw new Error('File transport runtime is not configured')
      }
      const { inbox, outbox } = await runtime.fileQueue.resolvePaths()
      await runtime.fileQueue.mkdir(inbox)
      await runtime.fileQueue.mkdir(outbox)
      return {
        responsePath: `${outbox}/${request.id}.response.json`,
        eventsPath: `${outbox}/${request.id}.events.jsonl`,
        seenEvents: 0,
      }
    },
    async sendRequest(connection: FileConnection, request: OpenAPIClientTransportRequest): Promise<void> {
      if (!runtime.fileQueue) {
        throw new Error('File transport runtime is not configured')
      }
      const { inbox } = await runtime.fileQueue.resolvePaths()
      await runtime.fileQueue.write(
        `${inbox}/${request.id}.json`,
        JSON.stringify({
          id: request.id,
          operationId: request.operationId,
          method: request.method,
          path: request.path,
          headers: runtime.stringifyHeaders(request.headers),
          input: request.params,
        }, null, 2),
      )
    },
    async getUpdate(connection: FileConnection, request: OpenAPIClientTransportRequest) {
      if (!runtime.fileQueue) return null
      while (true) {
        try {
          const eventsRaw = await runtime.fileQueue.read(connection.eventsPath)
          const lines = eventsRaw.split('\n').filter(Boolean)
          const line = lines[connection.seenEvents]
          if (line) {
            const payload = runtime.parseJson(line) as { id?: string; event?: any; data?: unknown }
            if (payload?.event) {
              connection.seenEvents += 1
              return { id: payload.id ?? request.id, event: payload.event, data: payload.data }
            }
          }
        } catch {}
        return null
      }
    },
    async getResult(connection: FileConnection, request: OpenAPIClientTransportRequest): Promise<OpenAPIClientTransportOutcome> {
      if (!runtime.fileQueue) {
        return { id: request.id, ok: false, error: new Error('File transport runtime is not configured') }
      }
      while (true) {
        const update = await this.getUpdate?.(connection, request)
        if (update) {
          await request.onEvent?.({ id: update.id, event: update.event as any, data: update.data })
          continue
        }
        try {
          const response = runtime.parseJson(await runtime.fileQueue.read(connection.responsePath)) as
            | { ok: true; result: unknown }
            | { ok: false; error?: { message?: string } }
          if (response?.ok) return { id: request.id, ok: true, result: response.result }
          if (response && response.ok === false) {
            return { id: request.id, ok: false, error: new Error(response.error?.message ?? 'File transport request failed') }
          }
        } catch (error) {
          if (!(error instanceof Error) || !/ENOENT/.test(error.message)) {
            throw error
          }
        }
        await runtime.delay(runtime.fileQueue.pollIntervalMs)
      }
    },
  }
}
